Change API type for `Tensor[]`: `ITensorList`.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #73641
* #72415
* __->__ #72414
* #73351
* #73350
* #69607
* #69606
* #68640
* #67964

Partially fixes: #66328

This PR makes the change to `ITensorList` for all other unstructured kernels. It changes the
signatures of every kernel that takes a `Tensor[]` as parameter.

ysiraichi committed:
Change API type for `Tensor[]`: `ITensorList`.

Partially fixes: #66328

This PR replaces `TensorList` by `ITensorList` in the dispatcher and C++ APIs. This change
makes jumping between the boxed and unboxed APIs cheaper, since there's no need to copy
tensors around anymore.

Summarizing, there are 3 types of modifications introduced by this PR:

- Codegen changes: generate signatures with `ITensorList` instead of `TensorList`
- Dispatcher chages: make it understand how to jump between the boxed and unboxed APIs
- Propagating changes: change the signatures of every C++ function whose operator that has
  `Tensor[]` as an argument type in its schema

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type for `Tensor[]`: `ITensorList`."


Partially fixes: #66328

This PR replaces `TensorList` by `ITensorList` in the dispatcher and C++ APIs. This change
makes jumping between the boxed and unboxed APIs cheaper, since there's no need to copy
tensors around anymore.

Summarizing, there are 3 types of modifications introduced by this PR:

- Codegen changes: generate signatures with `ITensorList` instead of `TensorList`
- Dispatcher chages: make it understand how to jump between the boxed and unboxed APIs
- Propagating changes: change the signatures of every C++ function whose operator that has
  `Tensor[]` as an argument type in its schema

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type for `Tensor[]`: `ITensorList`."


Partially fixes: #66328

This PR replaces `TensorList` by `ITensorList` in the dispatcher and C++ APIs. This change
allows functions not to restrict themselves to only one API. Thus, calling it becomes cheaper whenever
the APIs do not match (i.e. the function is unboxed, but the list is boxed), since there's no need to
copy tensors around anymore for that.

Summarizing, there are 3 types of modifications introduced by this PR:

- Codegen changes: generate signatures with `ITensorList` instead of `TensorList`
- Dispatcher chages: make it understand how to jump between the boxed and unboxed APIs
- Propagating changes: change the signatures of every C++ function whose operator that has
  `Tensor[]` as an argument type in its schema

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type for `Tensor[]`: `ITensorList`."


Partially fixes: #66328

This PR replaces `TensorList` by `ITensorList` in the dispatcher and C++ APIs. This change
allows functions not to restrict themselves to only one API. Thus, calling it becomes cheaper whenever
the APIs do not match (i.e. the function is unboxed, but the list is boxed), since there's no need to
copy tensors around anymore for that.

Summarizing, there are 3 types of modifications introduced by this PR:

- Codegen changes: generate signatures with `ITensorList` instead of `TensorList`
- Dispatcher chages: make it understand how to jump between the boxed and unboxed APIs
- Propagating changes: change the signatures of every C++ function whose operator that has
  `Tensor[]` as an argument type in its schema

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type for `Tensor[]`: `ITensorList`."


Partially fixes: #66328

This PR replaces `TensorList` by `ITensorList` in the dispatcher and C++ APIs. This change
allows functions not to restrict themselves to only one API. Thus, calling it becomes cheaper whenever
the APIs do not match (i.e. the function is unboxed, but the list is boxed), since there's no need to
copy tensors around anymore for that.

Summarizing, there are 3 types of modifications introduced by this PR:

- Codegen changes: generate signatures with `ITensorList` instead of `TensorList`
- Dispatcher chages: make it understand how to jump between the boxed and unboxed APIs
- Propagating changes: change the signatures of every C++ function whose operator that has
  `Tensor[]` as an argument type in its schema

[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -912,7 +913,7 @@ Tensor mm_batching_rule(const Tensor& self, const Tensor& other) {
 }
 
-Tensor cat_batching_rule(TensorList tensors, int64_t dim) {
+Tensor cat_batching_rule(const ITensorList& tensors, int64_t dim) {
```
just curious - is there a reason for the switch to const references?

I remember we needed it in `TensorBody.h` to avoid the #include cycles, but does it provide any benefit in the other places where we already take other reference types by value? (`TensorList`, `IntArrayRef`, etc)

bdhirsh replied:
It looks like `ArrayRef<Tensor>::front()` does bounds checking, while `operator[]` doesn't.

To preserve the existing error behavior, maybe we should add a `front()` method to `IList` that does bounds checking?

ysiraichi replied:
- We need them here, because `cat` was defined (e.g. `library.def("cat", ...)`) with the `const &`
- I believe they don't provide any benefits in this context

If we want to have non-const value `ITensorList`, I think we could add them when generating the method's signature. i.e. make the argument type for the dispatcher and C++ API for `Tensor[]` as `ITensorList` (no `const &`), and add them (the `const &`) when generating the method's signature in `TensorBody.h`.

bdhirsh replied:
ohhh yup. If we need them then SGTM

bdhirsh commented on aten/src/ATen/DeviceGuard.h:
```diff
@@ -26,12 +27,12 @@ inline optional<Device> device_of(const optional<Tensor>& t) {
+inline c10::optional<Device> device_of(ITensorList t) {
   if (!t.empty()) {
-    return device_of(t.front());
+    return device_of(t[0]);
```
It looks like `ArrayRef<Tensor>::front()` does bounds checking, while `operator[]` doesn't.

To preserve the existing error behavior, maybe we should add a `front()` method to `IList` that does bounds checking?

bdhirsh commented on aten/src/ATen/FunctionalTensorWrapper.cpp:
```diff
@@ -222,28 +222,14 @@ Tensor to_functional_tensor(const Tensor& tensor) {
 }
-c10::List<Tensor> to_functional_tensor(const c10::List<Tensor>& t_list) {
-  c10::List<Tensor> outputs;
+std::vector<Tensor> to_functional_tensor(ITensorList t_list) {
```
thanks for fixing :)

bdhirsh commented on aten/src/ATen/core/dispatch/DispatchKeyExtractor.h:
```diff
@@ -61,7 +61,7 @@ namespace detail {
       }
     }
-    void operator()(at::ArrayRef<at::Tensor> xs) {
+    void operator()(at::ITensorList xs) {
```
Hmm... does stuff break if we leave this as `ArrayRef<Tensor>`?

I might be wrong, but I think that this code lives entirely in unboxed world, which means that we know the underlying this is an `ArrayRef<Tensor>`. This is also pretty hot-path code in the dispatcher's runtime dispatch logic.

ysiraichi replied:
I think we could leave the `ArrayRef<Tensor>` implementation alongside the `ITensorList` one (not sure that was what you meant). That's because with the introduction of `ITensorList`, boxed lists of tensors might get to this code path.

bdhirsh replied:
I don't think that boxed lists of tensor would actually reach that code path - the `DispatchKeyExtractor` is responsible for merging all of the tensor-like arguments from an op into a single `DispatchKeySet`, and it has separate functions for doing that in the boxed world (`getDispatchKeySetBoxed()`) and the unboxed world (`getDispatchKeySetUnboxed`).

The boxed version is [here](https://github.com/pytorch/pytorch/blob/5e6f29661254e5ebc97fac3d829e5d455cda5864/aten/src/ATen/core/dispatch/DispatchKeyExtractor.h#L144), and it doesn't go through the `MultiDispatchKeySet` struct. So we can assume that only `ArrayRef<Tensor>` objects will ever get here (and I don't think we'll need an `ITensorList` template specialization.

ysiraichi replied:
Hmm. That was true. But, if I understand it correctly, `ITensorList` changes that. Consider, for example, the case where we somehow call `at::stack(boxed)` from the boxed world:

1. `boxed` (`IValue`) is converted to `List<Tensor>` (`ivalue_to_arg`), and then to `ITensorList` (let's call it `tlist`)
2. `at::stack` gets dispatched to `at::native::stack` function
3. `at::native::stack` calls `at::cat(tlist)`

In (3), we enter the dispatcher in the unboxed code path. However, one of its argumes is `tlist`, which is basically a pointer to a boxed representation `List<Tensor>`.

ysiraichi replied:
Hmmm. Good question. Maybe it's possible, but wouldn't that be the same as not having `ITensorList` at all, while cloning functions for `ArrayRef<Tensor>` and `List<Tensor>`? Anyway, I think the biggest challenge would be to duplicating every function that has a list of tensor as their parameters.

ysiraichi committed:
Update on "Change API type for `Tensor[]`: `ITensorList`."


Partially fixes: #66328

This PR replaces `TensorList` by `ITensorList` in the dispatcher and C++ APIs. This change
allows functions not to restrict themselves to only one API. Thus, calling it becomes cheaper whenever
the APIs do not match (i.e. the function is unboxed, but the list is boxed), since there's no need to
copy tensors around anymore for that.

Summarizing, there are 3 types of modifications introduced by this PR:

- Codegen changes: generate signatures with `ITensorList` instead of `TensorList`
- Dispatcher chages: make it understand how to jump between the boxed and unboxed APIs
- Propagating changes: change the signatures of every C++ function whose operator that has
  `Tensor[]` as an argument type in its schema

[ghstack-poisoned]

ysiraichi committed:
Fix flake8 errors. on "Change API type for `Tensor[]`: `ITensorList`."


Partially fixes: #66328

This PR makes the change to `ITensorList` for all other unstructured kernels. It changes the
signatures of every kernel that takes a `Tensor[]` as parameter.

[ghstack-poisoned]

ysiraichi committed:
Fix a few compiler errors. on "Change API type for `Tensor[]`: `ITensorList`."


Partially fixes: #66328

This PR makes the change to `ITensorList` for all other unstructured kernels. It changes the
signatures of every kernel that takes a `Tensor[]` as parameter.

[ghstack-poisoned]

ysiraichi committed:
Remove stale comment. on "Change API type for `Tensor[]`: `ITensorList`."


Partially fixes: #66328

This PR makes the change to `ITensorList` for all other unstructured kernels. It changes the
signatures of every kernel that takes a `Tensor[]` as parameter.

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type for `Tensor[]`: `ITensorList`."


Partially fixes: #66328

This PR makes the change to `ITensorList` for all other unstructured kernels. It changes the
signatures of every kernel that takes a `Tensor[]` as parameter.

[ghstack-poisoned]

ysiraichi committed:
Rebased to viable/strict. on "Change API type for `Tensor[]`: `ITensorList`."


Partially fixes: #66328

This PR makes the change to `ITensorList` for all other unstructured kernels. It changes the
signatures of every kernel that takes a `Tensor[]` as parameter.

[ghstack-poisoned]

ysiraichi committed:
Fix 'index.Tensor' test. on "Change API type for `Tensor[]`: `ITensorList`."


Partially fixes: #66328

This PR makes the change to `ITensorList` for all other unstructured kernels. It changes the
signatures of every kernel that takes a `Tensor[]` as parameter.

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type for `Tensor[]`: `ITensorList`."


Partially fixes: #66328

This PR makes the change to `ITensorList` for all other unstructured kernels. It changes the
signatures of every kernel that takes a `Tensor[]` as parameter.

[ghstack-poisoned]

ysiraichi committed:
Fix flake8 error. on "Change API type for `Tensor[]`: `ITensorList`."


Partially fixes: #66328

This PR makes the change to `ITensorList` for all other unstructured kernels. It changes the
signatures of every kernel that takes a `Tensor[]` as parameter.

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type for `Tensor[]`: `ITensorList`."


Partially fixes: #66328

This PR makes the change to `ITensorList` for all other unstructured kernels. It changes the
signatures of every kernel that takes a `Tensor[]` as parameter.

[ghstack-poisoned]

ysiraichi committed:
Invoke constructor explicitly when translating. on "Change API type for `Tensor[]`: `ITensorList`."


Partially fixes: #66328

This PR makes the change to `ITensorList` for all other unstructured kernels. It changes the
signatures of every kernel that takes a `Tensor[]` as parameter.

[ghstack-poisoned]

ysiraichi committed:
Rebased to 'viable/strict.' on "Change API type for `Tensor[]`: `ITensorList`."


Partially fixes: #66328

This PR makes the change to `ITensorList` for all other unstructured kernels. It changes the
signatures of every kernel that takes a `Tensor[]` as parameter.

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type for `Tensor[]`: `ITensorList`."


Partially fixes: #66328

This PR makes the change to `ITensorList` for all other unstructured kernels. It changes the
signatures of every kernel that takes a `Tensor[]` as parameter.

[ghstack-poisoned]

ysiraichi committed:
Rename 'ITensorList' to 'ITensorListRef'. on "Change API type for `Tensor[]`: `ITensorList`."


Partially fixes: #66328

This PR makes the change to `ITensorList` for all other unstructured kernels. It changes the
signatures of every kernel that takes a `Tensor[]` as parameter.

[ghstack-poisoned]

ysiraichi committed:
Unmerge `cat` and `Tensor?[]` support (mistakenly removed). on "Change API type for `Tensor[]`: `ITensorList`."


Partially fixes: #66328

This PR makes the change to `ITensorList` for all other unstructured kernels. It changes the
signatures of every kernel that takes a `Tensor[]` as parameter.

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type for `Tensor[]`: `ITensorList`."


Partially fixes: #66328

This PR makes the change to `ITensorList` for all other unstructured kernels. It changes the
signatures of every kernel that takes a `Tensor[]` as parameter.

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type for `Tensor[]`: `ITensorList`."


Partially fixes: #66328

This PR makes the change to `ITensorList` for all other unstructured kernels. It changes the
signatures of every kernel that takes a `Tensor[]` as parameter.

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type for `Tensor[]`: `ITensorList`."


Partially fixes: #66328

This PR makes the change to `ITensorList` for all other unstructured kernels. It changes the
signatures of every kernel that takes a `Tensor[]` as parameter.

[ghstack-poisoned]
